[REVIEW] Added tests for host_buffer and improved device_buffer and host_buffer implementation

jirikraus opened the pull request:
This pull request removes redundancies in the implementation of `device_buffer` and `host_buffer` and adds  the missing unit test for `host_buffer`. A unit test for `device_buffer` was not added, because the `device_buffer` implementation is fully covered by the `host_buffer` tests.

jirikraus commented:
@teju85 for his feedback.

teju85 commented on ml-prims/test/CMakeLists.txt:
```diff
@@ -79,6 +79,7 @@ add_executable(mlcommon_test
   transpose.cu
   unary_op.cu
   weighted_mean.cu
+  host_buffer.cu
```
minor knitpick! please keep the files in alphabetical order.

jirikraus replied:
Will fix it.

teju85 commented on ml-prims/src/common/buffer_base.hpp:
```diff
@@ -38,11 +43,23 @@ class buffer_base
+        if ( _capacity > 0 )
+        {
+            _data = static_cast<value_type*>(_allocator->allocate( _capacity*sizeof(value_type), _stream ));
+            CUDA_CHECK( cudaStreamSynchronize( _stream ) );
```
Depending on what we decide in #425 we might have to do this inside either the `host_buffer` or keep this here itself.

jirikraus replied:
Yes, but I see this as part of the discussion in #425 not this PR. So this is nothing to change in this PR now. Depending on how we conclude #425 a new PR might be needed. Agreed?

jirikraus replied:
@teju85 since your reacted with thumbs up on my comment I will resolve this conversation. Can you approve this PR or have I missed a change that you requested?

jirikraus commented:
@teju85 thanks for the review. I addressed your comments, rebased this change and applied the recently introduced code formatting rules. Can you take a look at this again and approve if this looks good to go?
Also tagging @cjnolet and @dantegd for visibility.

cjnolet commented on cpp/src_prims/common/buffer_base.hpp:
```diff
@@ -58,10 +78,49 @@ class buffer_base {
+      _data = new_data;
+      _capacity = new_capacity;
+    }
+  }
```
General CUDA memory question: Is it at all possible to shrink the array and let the underlying allocator know that the extra memory is available, since we have more control over that memory manager (RMM)?

jirikraus replied:
Interesting idea. I am not sure if RMM supports that. Anyway if I am not missing something it is not possible with the current allocator design in cuML. So I would say out of scope for this PR in any case. Do you agree?

cjnolet replied:
It's most definitely out of scope for this PR. It's a question I've been pondering myself and found this to be a good opportunity to ask.
